AX: AXTextOperation over multiple ranges is entirely broken

Sameesunkaria opened the pull request:
#### 4b8b62af6ff69eda42b461cbc3c46ec15428ca25
<pre>
AX: AXTextOperation over multiple ranges is entirely broken
<a href="https://bugs.webkit.org/show_bug.cgi?id=278928">https://bugs.webkit.org/show_bug.cgi?id=278928</a>
<a href="https://rdar.apple.com/problem/135021789">rdar://problem/135021789</a>

Reviewed by Tyler Wilcock.

While the AXTextOperation parameterized accessibility attribute provides an API for replacing
multiple ranges of text, in practice it would fail to apply changes after the first replacement in
many cases, especially when using smart replacement.

This was happening because after performing a replacement, in some cases, the remaining SimpleRanges
in `AccessibilityTextOperation` become orphaned as the node they were referring to would no longer
be in the tree. Specifically, this happened whenever the replacement was not performed using
`ReplaceSelectionCommand::performTrivialReplace` but instead followed the longer path in
`ReplaceSelectionCommand::doApply`.

To address this issue, before performing any replacements, we construct character ranges with
respect to their root editable element for all of the provided ranges. The root editable element
seems to be preserved in the tree after performing a replacement using the `ReplaceSelectionCommand`.
Then, we can reconstruct the `SimpleRange` for each replacement, ensuring we always use a valid
SimpleRange.

* LayoutTests/accessibility/mac/text-operation/text-operation-capitalize-expected.txt:
* LayoutTests/accessibility/mac/text-operation/text-operation-lowercase-expected.txt:
* LayoutTests/accessibility/mac/text-operation/text-operation-replace-across-multiple-fields-expected.txt: Added.
* LayoutTests/accessibility/mac/text-operation/text-operation-replace-across-multiple-fields.html: Added.

AXTextOperation can be provided text marker ranges from multiple editable fields on a page.
This test ensures that replacements are performed correctly across multiple fields.

* LayoutTests/accessibility/mac/text-operation/text-operation-replace-at-node-boundary-expected.txt: Added.
* LayoutTests/accessibility/mac/text-operation/text-operation-replace-at-node-boundary.html: Added.

Using ReplaceSelectionCommand to perform a replacement over a zero-length range (effectively
an insert) at node boundaries can result in one less whitespace than expected. To resolve
this, AccessibilityObject::performTextOperation now inserts text instead of replacing it when
the selection length is zero.

This test asserts on the correct behavior for that edge case.

* LayoutTests/accessibility/mac/text-operation/text-operation-smart-replace-expected.txt:
* LayoutTests/accessibility/mac/text-operation/text-operation-uppercase-expected.txt:
* LayoutTests/platform/mac-wk1/TestExpectations:
* LayoutTests/platform/mac-wk2/TestExpectations:
* LayoutTests/resources/accessibility-helper.js:
* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore::textOperationRangeFromRange):
(WebCore::rangeFromTextOperationRange):
(WebCore::AccessibilityObject::performTextOperation):

Canonical link: <a href="https://commits.webkit.org/284801@main">https://commits.webkit.org/284801@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/48c93518c3c1ebfed6c1789eeb07e5334a2e2719

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/70512 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/49918 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/23277 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/74601 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/21690 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/72629 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/57718 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/21530 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/55875 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/14344 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/73578 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/45400 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/60785 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/36336 "Found 1 new API test failure: /WPE/TestWebsiteData:/webkit/WebKitWebsiteData/service-worker-registrations (failure)") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/42059 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/18221 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/20051 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/63992 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/18570 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/76320 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/14738 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/17802 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/63594 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/14781 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/60852 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/63528 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15625 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/11572 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/5211 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/45720 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/487 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/46794 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/48071 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/46536 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Sameesunkaria commented on LayoutTests/resources/accessibility-helper.js:
```diff
@@ -303,7 +309,12 @@ async function selectElementTextById(id, axWebArea) {
+            return webArea.isTextMarkerRangeValid(selectedRange);
+        } else {
+            // isTextMarkerRangeValid is not available for WK1.
+            return webArea.textMarkerRangeLength(selectedRange) > 0;
```
Not really happy with this change. Perhaps it would be better to support `isTextMarkerRangeValid` for WK1 or skip the recently added `accessibility/mac/dynamic-selection.html` layout test for WK1, which depends on this helper function.

I'm open to making whichever change seems cleaner.

twilco replied:
I think we should skip the test on WK1 rather than add any branches to support it.

twilco commented on Source/WebCore/accessibility/AccessibilityObject.cpp:
```diff
@@ -1078,7 +1116,14 @@ Vector<String> AccessibilityObject::performTextOperation(const AccessibilityText
+            if (textOperationRange.characterRange.length)
+                frame->editor().replaceSelectionWithText(replacementString, Editor::SelectReplacement::Yes, operation.smartReplace == AccessibilityTextOperationSmartReplace::No ? Editor::SmartReplace::No : Editor::SmartReplace::Yes);
+            else
+                frame->editor().insertText(replacementString, nullptr);
```
It could be nice to have an inline comment documenting what this `nullptr` is, e.g. `frame->editor().insertText(replacementString, /* triggeringEvent */ nullptr);`

twilco commented on Source/WebCore/accessibility/AccessibilityObject.cpp:
```diff
@@ -1027,8 +1054,19 @@ Vector<String> AccessibilityObject::performTextOperation(const AccessibilityText
 
-    for (size_t i = 0; i < operation.textRanges.size(); ++i) {
-        auto textRange = operation.textRanges[i];
+    for (const auto& range : operation.textRanges) {
+        auto textOperationRange = textOperationRangeFromRange(range);
+        if (!textOperationRange) {
+            ASSERT_NOT_REACHED();
+            return result;
+        }
+
+        textOperationRanges.append(*textOperationRange);
+    }
```
Do we need two separate loops between this one and the `textOperationRanges` one? Could we just have one loop that iterates over `operation.textRanges` and calls `textOperationRangeFromRange` and `rangeFromTextOperationRange` as appropriate, and `continue`s if we get a `std::nullopt` from `textOperationRangeFromRange`? But maybe I'm missing some advantage to doing it this way.

Sameesunkaria replied:
Performing a replacement can invalidate other stored SimpleRanges, which prevents us from performing a selection, causing this bug.

That is why I am constructing the character ranges while the start/end containers of the stored SimpleRanges are still valid (i.e. **before** performing any replacement).

The call to `rootEditableElement` would fail for all orphaned range start/end containers if I attempted to construct the character range in the same loop as where I perform the replacement.

I tried to address this in the commit message:

>To address this issue, before performing any replacements, we construct character ranges with
>respect to their root editable element for all of the provided ranges. The root editable element
>seems to be preserved in the tree after performing a replacement using the `ReplaceSelectionCommand`.
>Then, we can reconstruct the `SimpleRange` for each replacement, ensuring we always use a valid
>SimpleRange.

twilco replied:
Ahh, right, thanks! Would you please add a comment explaining the reason for the separate loops so the reason is obvious directly in the code?

Sameesunkaria replied:
Added a comment in code :)

twilco commented on Source/WebCore/accessibility/AccessibilityObject.cpp:
```diff
@@ -1013,9 +1013,36 @@ Vector<SimpleRange> AccessibilityObject::findTextRanges(const AccessibilitySearc
+    textOperationRanges.reserveCapacity(operation.textRanges.size());
+
     Vector<String> result;
+    result.reserveCapacity(operation.textRanges.size());
```
It's slightly more efficient to use `reserveInitialCapacity` when sizing a vector for the first time.

twilco commented on Source/WebCore/accessibility/AccessibilityObject.cpp:
```diff
@@ -1013,9 +1013,36 @@ Vector<SimpleRange> AccessibilityObject::findTextRanges(const AccessibilitySearc
+
+static std::optional<TextOperationRange> textOperationRangeFromRange(const SimpleRange& range)
+{
+    auto* rootEditableElement = range.start.container->rootEditableElement();
```
Since `makeRangeSelectingNode` is a non-trivial operation, we should use a smart pointer instead of a raw pointer. `RefPtr` is probably fine here.

https://github.com/WebKit/WebKit/wiki/Safer-CPP-Guidelines#when-calling-a-non-trivial-member-function-on-an-object-hold-a-smart-pointer-to-the-object-on-the-stack

twilco commented on LayoutTests/resources/accessibility-helper.js:
```diff
@@ -318,19 +329,35 @@ async function selectPartialElementTextById(id, startIndex, endIndex, axWebArea)
-    range.setEnd(element.firstChild, endIndex);
-    selection.addRange(range);
+    if (element.setSelectionRange) {
+        // For textarea and input elements
```
WebKit coding style prefers comments to look like full sentences (so a period would be great in this comment and the next 🙂)

https://webkit.org/code-style-guidelines/#comments-sentences

twilco reviewed (commented):
Thanks for fixing this!

Sameesunkaria committed:
AX: AXTextOperation over multiple ranges is entirely broken
https://bugs.webkit.org/show_bug.cgi?id=278928
rdar://problem/135021789

Reviewed by Tyler Wilcock.

While the AXTextOperation parameterized accessibility attribute provides an API for replacing
multiple ranges of text, in practice it would fail to apply changes after the first replacement in
many cases, especially when using smart replacement.

This was happening because after performing a replacement, in some cases, the remaining SimpleRanges
in `AccessibilityTextOperation` become orphaned as the node they were referring to would no longer
be in the tree. Specifically, this happened whenever the replacement was not performed using
`ReplaceSelectionCommand::performTrivialReplace` but instead followed the longer path in
`ReplaceSelectionCommand::doApply`.

To address this issue, before performing any replacements, we construct character ranges with
respect to their root editable element for all of the provided ranges. The root editable element
seems to be preserved in the tree after performing a replacement using the `ReplaceSelectionCommand`.
Then, we can reconstruct the `SimpleRange` for each replacement, ensuring we always use a valid
SimpleRange.

* LayoutTests/accessibility/mac/text-operation/text-operation-capitalize-expected.txt:
* LayoutTests/accessibility/mac/text-operation/text-operation-lowercase-expected.txt:
* LayoutTests/accessibility/mac/text-operation/text-operation-replace-across-multiple-fields-expected.txt: Added.
* LayoutTests/accessibility/mac/text-operation/text-operation-replace-across-multiple-fields.html: Added.

AXTextOperation can be provided text marker ranges from multiple editable fields on a page.
This test ensures that replacements are performed correctly across multiple fields.

* LayoutTests/accessibility/mac/text-operation/text-operation-replace-at-node-boundary-expected.txt: Added.
* LayoutTests/accessibility/mac/text-operation/text-operation-replace-at-node-boundary.html: Added.

Using ReplaceSelectionCommand to perform a replacement over a zero-length range (effectively
an insert) at node boundaries can result in one less whitespace than expected. To resolve
this, AccessibilityObject::performTextOperation now inserts text instead of replacing it when
the selection length is zero.

This test asserts on the correct behavior for that edge case.

* LayoutTests/accessibility/mac/text-operation/text-operation-smart-replace-expected.txt:
* LayoutTests/accessibility/mac/text-operation/text-operation-uppercase-expected.txt:
* LayoutTests/platform/mac-wk1/TestExpectations:
* LayoutTests/platform/mac-wk2/TestExpectations:
* LayoutTests/resources/accessibility-helper.js:
* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore::textOperationRangeFromRange):
(WebCore::rangeFromTextOperationRange):
(WebCore::AccessibilityObject::performTextOperation):

Canonical link: https://commits.webkit.org/284801@main
